src/persist/src/storage.rs

```diff
@@ -72,40 +71,70 @@ pub trait Blob {
     fn set(&mut self, key: &str, value: Vec<u8>, allow_overwrite: bool) -> Result<(), Error>;
 }
 
-/// An implementation of [Persister] in terms of a [Buffer] and [Blob].
-pub struct StoragePersister<U: Buffer, L: Blob> {
+/// An implementation of a set of persisted streams in terms of a [Buffer] and [Blob].
+pub struct Persister<U: Buffer, L: Blob> {
     registered: HashSet<Id>,
     indexed: Arc<Mutex<Indexed<U, L>>>,
 }
 
-impl<U: Buffer, L: Blob> StoragePersister<U, L> {
-    /// Returns a [StoragePersister] initialized with previous data, if any.
+impl<U: Buffer, L: Blob> Persister<U, L> {
+    /// Returns a [Persister] initialized with previous data, if any.
     pub fn new(buf: U, blob: L) -> Result<Self, Error> {
-        Ok(StoragePersister {
+        Ok(Persister {
             registered: HashSet::new(),
             indexed: Arc::new(Mutex::new(Indexed::new(buf, blob)?)),
         })
     }
-}
-
-impl<U: Buffer, L: Blob> Persister for StoragePersister<U, L> {
-    type Write = StreamWriteHandle<U, L>;
 
-    type Meta = StreamMetaHandle<U, L>;
-
-    fn create_or_load(&mut self, id: Id) -> Result<Token<Self::Write, Self::Meta>, Error> {
+    /// Returns a token used to construct a persisted stream operator.
+    ///
+    /// If data was written by a previous [Persister] for this id, it's loaded and
+    /// replayed into the stream once constructed.
+    ///
+    /// Within a process, this can only be called once per id, even if that id
+    /// has since been destroyed. An `Err` is returned for calls after the
+    /// first. TODO: Is this restriction necessary/helpful?
+    pub fn create_or_load(
+        &mut self,
+        id: Id,
+    ) -> Result<Token<StreamWriteHandle<U, L>, StreamMetaHandle<U, L>>, Error> {
         if self.registered.contains(&id) {
             return Err(format!("internal error: {:?} already registered", id).into());
         }
         self.registered.insert(id);
+        {
+            let mut guard = self.indexed.lock()?;
+            guard.register(id);
+        }
         let write = StreamWriteHandle::new(id, self.indexed.clone());
         let meta = StreamMetaHandle::new(id, self.indexed.clone());
         Ok(Token { write, meta })
     }
 
-    fn destroy(&mut self, _id: Id) -> Result<(), Error> {
+    /// Remove the persisted stream.
+    ///
+    /// TODO: Should this live on Meta?
+    pub fn destroy(&mut self, _id: Id) -> Result<(), Error> {
         unimplemented!()
     }
+
+    /// Consume this [Persister] and return the underlying data
+    #[cfg(test)]
+    pub fn into_inner(self) -> Arc<Mutex<Indexed<U, L>>> {
```

Comment:
You're right, I don't think this is the right model anymore. I'm also not a big fan of using files in tests where it can easily be avoided. IMO the MemBlob and MemBuffer internals should get wrapped in `Arc<Mutex<>>` and shared so they're more like real impls.

Something like
```
struct MemBlobCore {
    lock: Option<String>,
    dataz: HashMap<String, Vec<u8>>
}
struct MemBlob {
    core: Arc<Mutex<MemBlobCore>>
}
```

Then there will be a central registry for getting them by "path" (the FnMut in blob_impl_test) so the data can be shared.
```
struct MemBlobRegistry {
    blobs_by_id: HashMap<Id, Arc<Mutex<MemBlobCore>>>,
}

impl MemBlobRegistry {
    fn open(&mut self, path: usize) -> MemBlob {
        ...
    }
}
```

The tests will then construct a Persister something like:
```
let buffer = ...
let blobs = MemBlobRegistry::new();
let persister = Persister::new(buffer.get(1), blobs.get(1))?;
...

// This is actually going to break right now because we haven't fixed the unlock stuff yet.
let persister_for_second_dataflow = Persister::new(buffer.get(1), blobs.get(1))?;
...
```

It'll be a bit of work, but these will be used in basically every test, so I think it'll be worth the investment.